### DISCUSSION TITLE
Board developement

### DISCUSSION THREAD

**Discussion by purewack** [micropython/micropython#16473] [Hardware & Peripherals]
I wanted to create my own board using an ESP32. 

I have already forked the repo and added my own board definitions along with my specific modules.
My intention is to develop more or less a stand alone product rather than a board that is used as a module - said board will include things like buttons, LCD, SD card slot, battery charger etc.

My question is: 
is Micropython the way to go for developing a product or perhaps Circuitpython? 
The implementation of my board will for example make extensive use of the ULP to monitor battery levels for example, as well as the board also displaying its REPL output on the LCD by default which could be controller via my custom module.

I don't know if this will be accepted as a 'board' if it was to be included in Micropython some day, as it is not like any other boards for Micropython.

Perhaps I should maintain my own fork and adapt it for my use only, but I would love to hear from people, and the maintainers of Micropython !

**Comment by aliarifat** [comment]
Circuitpython lacks ULP support. It may not be suitable for your battery monitoring and customization needs.  MicroPython supports the ESP32's Ultra-Low Power (ULP) coprocessor. This is convenient for battery monitoring. Since you are developing a standalone ESP32 board with displays and other peripherals, you can see this design for a reference: https://www.pcbway.com/blog/2/ESP32_Audio_Development_Board_f48936f7.html

**Reply by dhalbert** [reply]
> Circuitpython lacks ULP support.

CircuitPython _does_ support the ULP: https://docs.circuitpython.org/en/latest/shared-bindings/espulp/index.html

**Reply by purewack** [reply]
Ok then, suppose I went with Micropython - the board I'm making will be used as a sort of debugging tool for prototyping beard board circuits, making heavy use of SPI and gpio for low speed logic thru custom Python scripts. 

I'm trying to piece together the correct approach in my head as to not do anything stupid and upset the micropython community by not following conventions.

Im worried that by board may not be appropriate for being listed on the micropython page for downloads under the esp32 ports, as (to me) it feels more like a ready to go device rather than a generic board, given the custom scripts and start up behaviour which I am looking to implement.

Of course it would be more than possible for users to modify the scripts and what not for their own needs through mpremote (which is one of the better reasons I wanted to go with mp rather than hard coding a firmware in esp-idf)

I'm trying to be vague about what the board does exactly because it's still under development with only a few prototypes working, but I am developing a breadboard oriented device to speed up prototyping circuits, with the possibility of users writing their own bespoke scripts as they use the board over time.

**Reply by Josverl** [reply]
Perfect gift for this season 'beard board circuits' 🎅

**Comment by Josverl** [comment]
If you want to use MicroPython, you dont need to fork it.
You can setup your repo with your board config, hardware design, and all else you need.
And then use `git submodule` to add a reference to a specific MicroPython version or branch.

On disk this will place MicroPython in a subfolder of your repo, and your board files will need to accommodate for this.
